cpp/src/parquet/encoding-benchmark.cc

```diff
@@ -161,6 +176,268 @@ static void BM_DictDecodingInt64_literals(benchmark::State& state) {
+// Benchmark Decoding from Dictionary Encoding
+class BM_DictDecodingByteArray : public BenchmarkDecodeArrow {
+ public:
+  void DoEncodeData() override {
+    auto node = schema::ByteArray("name");
+    descr_ = std::unique_ptr<ColumnDescriptor>(new ColumnDescriptor(node, 0, 0));
+    auto encoder = MakeTypedEncoder<ByteArrayType>(Encoding::PLAIN,
+                                                   /*use_dictionary=*/true, descr_.get());
+    ASSERT_NO_THROW(encoder->Put(values_.data(), num_values_));
+    buffer_ = encoder->FlushValues();
+
+    auto dict_encoder = dynamic_cast<DictEncoder<ByteArrayType>*>(encoder.get());
+    ASSERT_NE(dict_encoder, nullptr);
+    dict_buffer_ =
+        AllocateBuffer(default_memory_pool(), dict_encoder->dict_encoded_size());
+    dict_encoder->WriteDict(dict_buffer_->mutable_data());
+    num_dict_entries_ = dict_encoder->num_entries();
+  }
+
+ protected:
+  std::unique_ptr<ColumnDescriptor> descr_;
+  std::unique_ptr<DictDecoder<ByteArrayType>> dict_decoder_;
+  std::shared_ptr<Buffer> dict_buffer_;
+  int num_dict_entries_;
+};
+
+BENCHMARK_DEFINE_F(BM_DictDecodingByteArray, DecodeArrow_Dense)
+(benchmark::State& state) {
+  for (auto _ : state) {
+    auto decoder = MakeTypedDecoder<ByteArrayType>(Encoding::PLAIN, descr_.get());
+    decoder->SetData(num_dict_entries_, dict_buffer_->data(),
+                     static_cast<int>(dict_buffer_->size()));
+    auto dict_decoder = MakeDictDecoder<ByteArrayType>(descr_.get());
+    dict_decoder->SetDict(decoder.get());
+    dict_decoder->SetData(num_values_, buffer_->data(),
+                          static_cast<int>(buffer_->size()));
+    auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder.get());
+
+    ::arrow::internal::ChunkedBinaryBuilder builder(
+        static_cast<int>(dict_buffer_->size()), ::arrow::default_memory_pool());
+    byte_array_decoder->DecodeArrow(num_values_, 0, valid_bits_, 0, &builder);
+  }
+
+  state.SetBytesProcessed(state.iterations() * total_size_);
+}
+
+BENCHMARK_REGISTER_F(BM_DictDecodingByteArray, DecodeArrow_Dense)
+    ->Range(MIN_RANGE, MAX_RANGE);
+
+BENCHMARK_DEFINE_F(BM_DictDecodingByteArray, DecodeArrowNonNull_Dense)
+(benchmark::State& state) {
+  for (auto _ : state) {
+    auto decoder = MakeTypedDecoder<ByteArrayType>(Encoding::PLAIN, descr_.get());
+    decoder->SetData(num_dict_entries_, dict_buffer_->data(),
+                     static_cast<int>(dict_buffer_->size()));
+    auto dict_decoder = MakeDictDecoder<ByteArrayType>(descr_.get());
+    dict_decoder->SetDict(decoder.get());
+    dict_decoder->SetData(num_values_, buffer_->data(),
+                          static_cast<int>(buffer_->size()));
+    auto byte_array_decoder = dynamic_cast<ByteArrayDecoder*>(dict_decoder.get());
+
+    ::arrow::internal::ChunkedBinaryBuilder builder(
+        static_cast<int>(dict_buffer_->size()), ::arrow::default_memory_pool());
+    byte_array_decoder->DecodeArrowNonNull(num_values_, &builder);
```

Comment:
This benchmark differs on only one line from the previous one, can you refactor to reuse code?

https://www.diffchecker.com/VPUwaQ0t